carma_wm_ctrl/src/WMBroadcaster.cpp

```diff
@@ -129,26 +141,57 @@ lanelet::ConstLaneletOrAreas WMBroadcaster::getAffectedLaneletOrAreas(const cav_
   PJ* geofence_in_map_proj = proj_create_crs_to_crs(PJ_DEFAULT_CTX, geofence_msg.proj.c_str(), base_map_georef_.c_str(), NULL);
   
   // convert all geofence points into our map's frame
-  std::vector<lanelet::Point3d> gf_pts_in_base_map;
+  std::vector<lanelet::Point3d> gf_pts;
   for (auto pt : geofence_msg.points)
   {
     PJ_COORD c{{pt.x, pt.y, pt.z, 0}};
     PJ_COORD c_out;
     c_out = proj_trans(geofence_in_map_proj, PJ_FWD, c);
-    gf_pts_in_base_map.push_back(lanelet::Point3d{base_map_->pointLayer.uniqueId(), c_out.xyz.x, c_out.xyz.y});
+    gf_pts.push_back(lanelet::Point3d{current_map_->pointLayer.uniqueId(), c_out.xyz.x, c_out.xyz.y});
   }
 
   // Logic to detect which part is affected
-  int nearest_element_num = 5;
+  double lane_max_width = 4;
+  
   std::unordered_set<lanelet::Lanelet> affected_lanelets;
-  for (lanelet::Point3d pt : gf_pts_in_base_map)
+  for (int idx = 0; idx < gf_pts.size(); idx ++)
   {
-    // get nearest few lanelets as checking every single llt will be slower
-    // and check if the point is within that llt
-    for (lanelet::Lanelet llt : base_map_->laneletLayer.nearest(pt, nearest_element_num))
+    std::unordered_set<lanelet::Lanelet> possible_lanelets;
+    // get nearest few nearest llts within lane_max_width
+    // which actually house this geofence_point
+    auto searchFunc = [&](const lanelet::BoundingBox2d& lltBox, const lanelet::Lanelet& llt) 
+    {
+      bool should_stop = boost::geometry::distance(gf_pts[idx].basicPoint2d(), llt.polygon2d()) > lane_max_width;
+      if (!should_stop && boost::geometry::within(gf_pts[idx].basicPoint2d(), llt.polygon2d()))
+      {
+        possible_lanelets.insert(llt);
+      }
+      return should_stop;
+    };
+   
+    current_map_->laneletLayer.nearestUntil(gf_pts[idx], searchFunc);
+
+    // among these llts, filter the ones that are on same direction as the geofence
+    if (idx + 1 == gf_pts.size()) // break if this is the last gf_pt after saving everything
+    {
+      for (auto llt: possible_lanelets)
+      {
+        affected_lanelets.insert(llt);
+      }
+      break;
+    } 
+  
+    // check if each lines connecting end points of the llt is crossing with the line connecting current and next gf_pts
+    for (auto llt: possible_lanelets)
     {
-      if (boost::geometry::within(pt.basicPoint2d(), llt.polygon2d()))
+    
+      lanelet::BasicLineString2d gf_dir_line({gf_pts[idx].basicPoint2d(), gf_pts[idx+1].basicPoint2d()});
+      lanelet::BasicLineString2d llt_boundary({(llt.leftBound2d().end() -1)->basicPoint2d(), (llt.rightBound2d().end() - 1)->basicPoint2d()});
+      if (boost::geometry::intersects(llt_boundary, gf_dir_line))
+      {
+        // record the llts that are on the same dir
         affected_lanelets.insert(llt);
+      }
```

Comment:
This looks like a good implementation of your original algorithm, but I think there is an edge case for crossing lanelets. Lets discuss directly.